edb/edgeql/compiler/__init__.py

```diff
@@ -128,55 +128,61 @@
 from __future__ import annotations
 from typing import *
 
-from edb import errors
-
-from edb.edgeql import parser as ql_parser
+# WARNING: this package is in a tight import loop with various modules
+# in edb.schema, so no direct imports from either this package or
+# edb.schema are allowed at the top-level.  If absolutely necessary,
+# use the lazy-loading mechanism.
 
-from edb.common import debug
-from edb.common import markup  # NOQA
+import functools
 
-from edb.schema import functions as s_func
-from edb.schema import types as s_types
+from edb import errors
 
 from edb.edgeql import ast as qlast
-from edb.ir import ast as irast
-from edb.ir import staeval as ireval
+from edb.edgeql import parser as qlparser
 
-from . import dispatch
-from . import inference
-from . import setgen
-from . import stmtctx
+from edb.common import debug
 
-from . import expr as _expr_compiler  # NOQA
-from . import config as _config_compiler  # NOQA
-from . import stmt as _stmt_compiler  # NOQA
+from .options import CompilerOptions as CompilerOptions  # "as" for reexport
 
 if TYPE_CHECKING:
-    from edb.schema import name as s_name
-    from edb.schema import objects as s_obj
     from edb.schema import schema as s_schema
+    from edb.schema import types as s_types
+
+    from edb.ir import ast as irast
+    from edb.ir import staeval as ireval
+
+    from . import dispatch as dispatch_mod
+    from . import inference as inference_mod
+    from . import stmtctx as stmtctx_mod
+else:
+    # Modules will be loaded lazily in _load().
+    dispatch_mod = None
+    inference_mod = None
+    irast = None
+    ireval = None
+    stmtctx_mod = None
+
 
+#: Compiler modules lazy-load guard.
+_LOADED = False
 
+
+def compiler_entrypoint(func: Callable[..., Any]) -> Callable[..., Any]:
```

Comment:
Alternatively you can use a module-level `__getattr__`

Reply:
I think that would be even cleaner.

Reply:
I don't think `__getattr__` would help here, since we need to populate the local namespace in the module itself. It would only work if the internals were imported via a proxy module with `__getattr__` which doesn't seem cleaner.